Fix random element selection for large hash tables.

gfemec opened the pull request:
When a database on a 64 bit build grows past 2^31 keys, the underlying hash table expands to 2^32 buckets. After this point, the algorithms for selecting random elements only return elements from half of the available buckets because they use `random()` which has a range of 0 to 2^31 - 1. This causes problems for eviction policies which use `dictGetSomeKeys` or `dictGetRandomKey`. Over time they cause the hash table to become unbalanced because, while new keys are spread out evenly across all buckets, evictions come from only half of the available buckets. Eventually this half of the table starts to run out of keys and it takes longer and longer to find candidates for eviction. This continues until no more evictions can happen. This solution addresses this by using a 64 bit PRNG when needed.

### Performance
To test the performance impact of this change, we used a LUA script to rapidly insert keys into databases at max memory using the affected eviction policies and compared results with and without this change. There were no meaningful differences with all databases achieving around 100k evictions per second.

### Test Coverage
While the basic eviction functionality is covered by existing tests, there isn't one that covers what happens after 2^31 keys. Because of the time and memory required to run a test like this I did not add one.

madolson commented:
Redis officially supports 2^32 keys, so this seems like something that should be addressed if we don't randomly evict keys in that space. So I know that this will actually produce a result that is less random than generating a pure 62 bit PRN, since the numbers produced are cyclical. Wondering if it would be slightly more correct to implement a proper 62 bit long random solution (Like http://www.math.sci.hiroshima-u.ac.jp/m-mat/MT/VERSIONS/C-LANG/mt19937-64.c) and then have a header that defines which random we used based on the size of the system. 

(Also, team terran!)

gfemec commented:
Thanks for taking a look! Using a 64 bit PRNG sounds like a great idea to me. I've updated the PR to use this and reran our tests. As expected it does fix the problem of eviction stalling after 2^31 keys and there doesn't appear to be a negative performance impact. The PRNG uses a couple KB of state and the license is BSD.

oranagra commented:
@madolson where does it say that redis supports 2^32 keys? i remember a few commits recently to dict to make it 64 bit compatible.
I'm not sure how i feel about the use of an "external" random generator (kinda liked the previous solution better). @yossigo WDYT?
if we keep it, we probably need to call the init function in redis-cli, redis-benchmark, redis-check-rdb, etc. these don't currently use the random features of dict.c, but some day they might.

madolson commented:
https://redis.io/topics/faq
@oranagra

madolson commented:
My original thought for keeping it is that I think it would be generally useful to have a 64 bit random number generator. It makes the rest of the code easier to understand, and will make sure we don't accidentally forgot to not add more bits again. 

Btw, I would be totally okay just using the random built-in to concatenate a 62 bit random number, there were just concerns about performance, so I proposed another option. 

We should probably seed redis-benchmark. Not sure how necessary it is to seed the other two.

gfemec commented:
I added code to seed the new PRNG in `redis-benchmark`. Let me know if I should add it to `redis-cli` and `redis-check-rdb` as well. Note that currently they aren't seeding `random()` so I would add calls for both RNGs.

For reference, the original approach of using multiple calls to `random()` is here: redis@628187f0e129192f6d6fe35bafbbd309a1fe8c94

gfemec commented:
I'd be happy to provide any additional information that might help us make a decision here. Perhaps I could add `dictGetRandomKey` to `dict-benchmark` and show results before/after each potential solution? I can see advantages to using a proper 64 bit PRNG (better "randomness" due to a longer period, more maintainable code, etc.) but understand there are drawbacks to introducing a dependency. Is there anyone else who we should ask for an opinion?

oranagra commented:
@gfemec sorry for the delay (busy days)...
i suppose such a benchmark can help.
e.g. if it'll show a performance boost that we'll be unable to ignore, there's a chance it'll force us to accept the dependency.
if it won't, then we may be still stuck on making a decision.

i wonder if @filipecosta90 or @yossigo have any opinion on that.
i.e. choosing between calling libc `random()` twice, and PRNG.

gfemec commented:
No problem, thanks for following up. I added `dictGetRandomKey` to `dict-benchmark` and rebased the branch on unstable to resolve merge conflicts. Here are the results of running `dict-benchmark` compiled for a 64 bit architecture on my machine. One thing to mention, the glibc implementation of `random()` uses a mutex to synchronize access making it thread-safe. The 64-bit PRNG library didn't come with a similar locking feature. For comparison, I also ran a test where I added a mutex to the 64-bit PRNG to make it thread-safe (the last row).

| Version  | Average Milliseconds Per 5M Calls (5 samples) |
| - | - |
| original | 1224 |
| `random()` (< 2^31 buckets) | 1258 |
| `random()` (>= 2^31 buckets) | 1430 |
| mt19937 64-bit PRNG | 1019 |
|  mt19937 64-bit PRNG (thread-safe) | 1319 |

In summary, if we want thread safety, it looks like the solution which uses `random()` is faster for the case where we have fewer than 2^31 buckets but slower if we have more.

madolson commented:
Yeah, re:invent makes everything so busy. So it seems like it's about 40ns faster. The computation part of "randomkey" usually takes 3-5us on my machine, that comes out to about a 1% difference. (Roughly)

I still like randomULong() for clarity, and I would be okay paying the cost of always doing random() twice so we don't have the extra branching. Since we have both options now, I guess everyone can weigh in and we can pick the right path.

oranagra commented:
we don't use random that heavily, so it seems to me the small difference in performance is not the factor for decision.
we also don't really use random from different threads at the moment, but we might some day, so i'm a bit worried about using a non-thread safe version.
I think the decision should probably based on cleaner code, however each version has a clean aspect from a different angle (one is nicer in the usage, but adds a dependency).
let's wait for more feedback (added it to the 6.2 project so that we don't forget)

yossigo commented:
I lean towards the new implementation. While we don't rely on random heavily, it's still nice to have a faster one (and we can work around mutexes with thread-local storage or create explicit context structs) and never look back.

madolson commented:
@redis/redis-contributors Other thoughts?

madolson reviewed (dismissed):
We reviewed it again and decided to accept it with the updated code. Oran will also take a look, but LGTM. Appreciate your patience!

guybe7 commented:
what about dropping the PRNG and do something like:
```
#if ULONG_MAX >= 0xffffffffffffffff
#define randomULong() ((unsigned long) random()) << 32 | random()
#else
#define randomULong() random()
#endif
```

oranagra commented:
@guybe7 that was sorta what was done in the original version of this PR.
mentioned as `random() (>= 2^31 buckets)` in the above benchmark.
the compromise here is about having another dependency, vs slightly slower code.
we ended up deciding to keep the dependency.

oranagra commented:
@gfemec one last change i'd like to request and then we'll merge this.
i think redis-cli and redis-check-rdb must call init_genrand64 too. so that when some day they'll start using random features of dict, it'll not be buggy.

IIUC, it's not just seeding the random, but also properly initializing this "unit".
without initialization, maybe the random values are poor or buggy.
so i don't care if we seed it with a constant 0, or time^pid, but i do want to init it.
that said, i don't see any reason not to seed it the same way we do for other executables (with time and pid).

gfemec commented:
I added a commit to initialize the 64-bit PRNG in `redis-cli` and `redis-check-rdb`.

Just so there's no confusion, as part of generating a random number, the 64-bit PRNG [checks](https://github.com/redis/redis/pull/8133/files#diff-c5e468806a7a6e67bc0b4050344438986130fc7735f295b2da826924b55cedb3R116) if it needs to be initialized and, if so, will initialize itself with a fixed seed. So, if we start using it and forget the initialization in another binary (like `redis-check-aof`), the behavior will still be well-defined.

oranagra commented:
ohh, that's good to hear (didn't bother to read the PRNG code).
this reminds me that we considered adding a mutex to make it thread-safe.
this isn't currently a problem since the only place that uses random and threads (without the GIL locked) is the benchmark, which doesn't use the dict random feature at the moment.
